go-controller/pkg/clustermanager/egressip_event_handler.go

```diff
@@ -133,16 +132,12 @@ func (h *egressIPClusterControllerEventHandler) UpdateResource(oldObj, newObj in
 			}
 			return nil
 		}
-		if isOldReady == isNewReady && !isHostCIDRsAltered {
```

Comment:
so walking through the consequences of not skipping potentially duplicate events:

- Delete path: 

1. We list all egress ips, with kube client (bad and we should fix that), negligible impact
2. If we already deleted this node, none of the status will match so no reconcile, no perf impact

- Add path:
1. Set node reachable, takes a global allocator lock, we could improve this with per key map lock, but shouldnt be a big deal since 15 threads competing for the lock on an operation that takes no time
2. Add egress node: we get egress ips via kube client again (wth...), then we go through every egress IP and reconcile...yikes

This seems like a pretty big performance issue if there were a lot of egress IPs. We would constantly be churning on every node update all egress ips. Or am I missing something?

Would a better approach be to expose the reAddorDelete map that the reachability check uses, and then we can have the state if the node wasReachable. Then we can change this if statement to be:

if isOldReady == isNewReady && !isHostCIDRsAltered && wasReachable == isReachable ?

Reply:
```
Would a better approach be to expose the reAddorDelete map that the reachability check uses, and then we can have the state if the node wasReachable. Then we can change this if statement to be:
```
I had the same thought to have a previous state of Reachability but the problem here is that the kubelet updates we get with oldNode, newNode both those obejcts will have healthy Reachability matrix so it still doesn't take action :( and doesn't help here.

Reply:
```
This seems like a pretty big performance issue if there were a lot of egress IPs. We would constantly be churning on every node update all egress ips. Or am I missing something?
```
SO this is a legit scale question, one I had too. But the thing to notice inside the add/delete node func is that we call the reconcileEIP only when `status!=spec`; so only *if the EIP needs to be reconciled we are calling reconcile* not otherwise.

Reply:
yeah true for delete its what you said, will have no impact because status.Node won't be that node..or at least it would/should get removed.

Reply:
> But the thing to notice inside the add/delete node func is that we call the reconcileEIP only when status!=spec; so only if the EIP needs to be reconciled we are calling reconcile not otherwise; I meant: https://github.com/ovn-org/ovn-kubernetes/pull/4102/files#diff-f706ecf5fbc17ede0f5a27577b1d4b3c08c00fe2837f6c1be461503de71ce21bR760 for the update/add case.

That also means that we will always reconcile egressIPs that cannot be assigned for whatever reason.

Reply:
> > But the thing to notice inside the add/delete node func is that we call the reconcileEIP only when status!=spec; so only if the EIP needs to be reconciled we are calling reconcile not otherwise; I meant: https://github.com/ovn-org/ovn-kubernetes/pull/4102/files#diff-f706ecf5fbc17ede0f5a27577b1d4b3c08c00fe2837f6c1be461503de71ce21bR760 for the update/add case.
> 
> That also means that we will always reconcile egressIPs that cannot be assigned for whatever reason.

yes agreed this will be a bit annoying, but given controllers are supposed to always try to reconcile things its perhaps the expected scenario in case there are eips that are unassigned - maybe it will prompt users to fix configs better.

Reply:
Hard for me to remember all of the context to this, but looking at my comments I think maybe I was confusing health check state with "was this thing programmed in CPIPC" state. I still think this PR in its current state is not a good solution.

It looks like EgressNode Add relies on CloudPrivateIPConfigType. If the problem is that the latter object is not programmed in time, and egress node times out and future updates to do not signal egress node to happen again, then imo the solution is to make CloudPrivateIPConfigType requeue affected egress node objects when it finally succeeds. Kind of like k8s endpoint events can cause services to reconcile again.

Similarly if an egress node event happens that relies on CloudPrivateIPConfigType, it can query  that controller's cache to see if it has been added correctly. If not, requeue the CloudPrivateIPConfigType and try again later.

As for healthchecks, I hope Jaime's refactor will force requeue of objects when healthcheck changes, including egressNode and CloudPrivateIPConfigType @jcaamano fyi

Reply:
A long term solution is level driven controllers as hinted in the description and discussed in meetings in the past, this is also why there was a card for refactoring EIP into level driven which didn't happen yet (every other object should simply add eip key and there should be 1 reconcile function), we need a backportable solution that doesn't involve major changes like requeueing nodes for cpipc OR requeueing cpipc for nodes etc.

Reply:
> 
> It looks like EgressNode Add relies on CloudPrivateIPConfigType. If the problem is that the latter object is not programmed in time, and egress node times out and future updates to do not signal egress node to happen again, then imo the solution is to make CloudPrivateIPConfigType requeue affected egress node objects when it finally succeeds. Kind of like k8s endpoint events can cause services to reconcile again.

no that's not it, CPIPC's failure does trigger a retry add for the node, but when kubelet update event comes the node retry is over written; its a timing thing + event drivenness

> 
> Similarly if an egress node event happens that relies on CloudPrivateIPConfigType, it can query that controller's cache to see if it has been added correctly. If not, requeue the CloudPrivateIPConfigType and try again later.

technically speaking we should all be adding only EIP key to the queue for any object changes..

> 
> As for healthchecks, I hope Jaime's refactor will force requeue of objects when healthcheck changes, including egressNode and CloudPrivateIPConfigType @jcaamano fyi

I need a backportable solution for this.